Add Axes.ecdf() method.

anntzer opened the pull request:
## PR Summary

See discussion at #16561 (attn @Wrzlprmft).  I chose to implement remove_redundant (under the name "compress") as I can see cases where it would be helpful for performance, but left "absolute" ecdfs unimplemented, as I have never seen them (and https://stats.stackexchange.com/questions/451601/what-are-absolute-ecdfs-called-if-anybody-uses-them didn't attract too much activity).

I updated the histogram_cumulative example to showcase this (as one should basically always use ecdf() instead of hist(..., cumulative=True, density=True), but let's not overdo it and immediately consider removing that, as it's probably extremely widely used).  Note that I removed the reference to astropy's bin selection examples, as it's already mentioned elsewhere (e.g. the histogram_features example) and also doesn't really apply as is to cumulative histograms.

---

Note: I am not overly convinced by silently dropping nans; we could also error on them.  (+/-inf should clearly be supported, as they have a non-ambiguous interpretation, which I have relied on before).

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

oscargus commented:
`.. added`...?

jklymak commented:
This is basically just `ax.plot(np.sort(x), np.arange(len(x)))`?   Even with weights and normalizations, this seems pretty trivial for users to compute themselves, and I'm not a huge fan of providing folks with statistical black boxes in Matplotlib core.

anntzer commented:
@oscargus @story645 I have addressed your comments.

@jklymak This was already argued at https://github.com/matplotlib/matplotlib/issues/16561#issuecomment-589282916 and https://github.com/matplotlib/matplotlib/issues/16561#issuecomment-591026865: the tricky part is not the calculation (which is indeed pretty trivial), but in selecting the correct drawstyle and adding the correct point at the correct end so that the ecdf is indeed a *step* plot (see https://en.wikipedia.org/wiki/Empirical_distribution_function, https://www.itl.nist.gov/div898/software/dataplot/refman1/auxillar/ecdfplot.htm, https://fr.mathworks.com/help/stats/ecdf.html, ...) instead of connecting the points with diagonal lines (which is actually wrong).

anntzer commented:
attn @Phlya @jaroslawr @mwaskom who have also commented on the original thread.

jaroslawr commented:
I think this is great, histograms, box plots and violin plots are already in matplotlib and ECDFs are a broadly useful type of plot in many different application areas. I would probably use pure matplotlib if only it had this feature. Libraries like seaborn come with the cost of a whole abstraction layer over matplotlib - you often have to understand both seaborn and matplotlib and how seaborn passes arguments to matplotlib etc. Filling in a few gaps like this PR does would make using pure matplotlib much more attractive.

oscargus commented on doc/api/axes_api.rst:
```diff
@@ -111,9 +111,10 @@ Statistics
 
    Axes.boxplot
    Axes.violinplot
+   Axes.ecdf
```
```suggestion
   Axes.ecdf
   Axes.violinplot
```
? (As you ordered the two remaining ones...)

anntzer replied:
sure

oscargus commented on doc/users/next_whats_new/ecdf.rst:
```diff
@@ -0,0 +1,4 @@
+``Axes.ecdf``
+~~~~~~~~~~~~~
+A new Axes method, `~.Axes.ecdf`, allows plotting empirical cumulative
+distribution functions without any binning.
```
I'd suggest adding an example plot here to make it a bit more visible. (It is not every day that we add a completely new plot type...)

Subject to getting general approval of the idea etc.

anntzer replied:
sure

Wrzlprmft commented:
Would it make sense to warn the users who are still using `hist(...,cumulative=True)`? At the very least, the documentation should point to the new function.

oscargus commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7135,6 +7135,99 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+        **kwargs
+            Keyword arguments control the `.Line2D` properties:
+
+            %(Line2D:kwdoc)s
```
```suggestion
            %(Line2D:kwdoc)%
```

anntzer replied:
no?

oscargus replied:
I may be wrong, but the current one is also wrong (or something else is wrong). Check the outcome.

anntzer replied:
ah, fixed

anntzer commented:
@Wrzlprmft While I agree with you that there's rather few use cases of hist(..., cumulative=True) that are not better served than by ecdf(), let's first get the method in and decide on whether to include your warning later; I don't want to be derailed into a side-discussion.

anntzer commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7135,6 +7135,100 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+            else:
+                line, = self.plot([1, *1 - cweights], [*x, x[-1]],
+                                  drawstyle="steps-post", **kwargs)
+            print(line.get_data())
```
woops, fixed.

anntzer commented:
General agreement during call was to error on nans instead, and also error on any input that has masked values.  Also modify the docstring to suggest various ways to handle nans (ignore them, map them to +/-inf).

jklymak commented:
Draft until above is implemented

anntzer commented:
Done.

jklymak commented:
@oscargus did you have a chance to take a second look at this?

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+        """
+        Compute and plot the empirical cumulative distribution function of *x*.
+
+        .. versionadded:: 3.7
```
This needs updating to 3.8

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+
+        Returns
+        -------
+        Line2D
```
Link this?

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+        x = x[argsort]
+        if weights is None:
+            # Ensure that we end at exactly 1, avoiding floating point errors.
+            cweights = (1 + np.arange(len(x))) / len(x)
```
I'm a bit confused as to what `cweights` is here? perhaps a comment explaining what it is would be good?

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+            # Ensure that we end at exactly 1, avoiding floating point errors.
+            cweights = (1 + np.arange(len(x))) / len(x)
+        else:
+            weights = np.take(weights, argsort)
```
```suggestion
            # Sort weights
            weights = np.take(weights, argsort)
```

dstansby commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7136,6 +7136,107 @@ def hist2d(self, x, y, bins=10, range=None, density=False, weights=None,
+            weights = np.take(weights, argsort)
+            cweights = np.cumsum(weights / np.sum(weights))
+        if compress:
+            compress_idxs = [0, *(x[:-1] != x[1:]).nonzero()[0] + 1]
```
```suggestion
            # Get indices of unique values in x
            compress_idxs = [0, *(x[:-1] != x[1:]).nonzero()[0] + 1]
```

dstansby reviewed (approved):
Looks good to me - I have some minor suggestions, and the version change the docs is blocking, but feel free to self-merge when that's done

anntzer commented:
Thanks, I've handled all the comments.

tacaswell commented on doc/users/next_whats_new/ecdf.rst:
```diff
@@ -0,0 +1,12 @@
+A new Axes method, `~.Axes.ecdf`, allows plotting empirical cumulative
+distribution functions without any binning.
+
+.. plot::
```
```suggestion
.. plot::
    :include-source: true
```

or we just get the image without the source!

anntzer replied:
done
